accounting-service/app/src/main/java/com/microservices/accounting/service/PaymentServiceImpl.java

```diff
@@ -0,0 +1,84 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.servlet.UnavailableException;
+import java.util.Objects;
+import java.util.Random;
+
+@Service
+public class PaymentServiceImpl implements PaymentService {
+    private PaymentRepository paymentRepository;
+    private final Logger logger = LoggerFactory.getLogger(PaymentServiceImpl.class);
+
+    @Autowired
+    public PaymentServiceImpl(PaymentRepository paymentRepository) {
+        this.paymentRepository = paymentRepository;
+    }
+
+    @Override
+    public PaymentDetailsDto invokePayment(InvokePaymentDto invokePaymentDto) {
+        Objects.requireNonNull(invokePaymentDto);
+
+        logger.info("Invoking payment: {}...", invokePaymentDto);
+        CardInfo cardInfo = invokePaymentDto.getUser().getCardInfo();
+        PaymentStatus paymentStatus = (cardInfo == CardInfo.VALID && isRequestApproved())
+                ? PaymentStatus.ACCEPTED
+                : PaymentStatus.DENIED;
+
+        PaymentEntity savedPayment = paymentRepository.save(new PaymentEntity(invokePaymentDto, paymentStatus));
+        logger.info("Saved payment: {}", savedPayment);
+        return new PaymentDetailsDto(savedPayment);
+    }
+
+    @Override
+    @Transactional
+    public PaymentDetailsDto revertPayment(RevertPaymentDto revertPaymentDto) throws UnavailableException {
+        Objects.requireNonNull(revertPaymentDto);
+        logger.info("Reverting payment: {}", revertPaymentDto);
+
+        PaymentEntity paymentToRevert = paymentRepository.findById(revertPaymentDto.getPaymentId())
+                .orElseThrow(() -> new IllegalArgumentException("No payment with id " + revertPaymentDto.getPaymentId()));
+
+        switch (paymentToRevert.getPaymentStatus()) {
+            case REVERTED:
+                logger.info("Payment with id {} is already reverted", paymentToRevert.getPaymentId());
+                return new PaymentDetailsDto(paymentToRevert);
+            case DENIED:
+                logger.error("Payment with id {} cannot be reverted. Has a denied status", paymentToRevert.getPaymentId());
+                throw new IllegalArgumentException("Denied payment cannot be reverted");
+            case ACCEPTED:
+                if (isRequestApproved()) {
+                    paymentToRevert.setPaymentStatus(PaymentStatus.REVERTED);
+                    PaymentEntity updatedPaymentEntity = paymentRepository.save(paymentToRevert);
+                    logger.info("Payment with id {} is reverted");
+                    return new PaymentDetailsDto(updatedPaymentEntity);
+                } else {
+                    logger.error("Payment reversion cannot be processed");
+                    throw new UnavailableException("Payment reversion cannot be processed now. Try later");
```

Comment:
What is the new exception?

Reply:
If the reversion cannot be performed, but revertDto is correct, the only reason for failing (in my opinion) is service unavailibility. What should be thrown in your opinion?

Reply:
IllegalStateException till we don't have our own